File: tswow-scripts/compile/MPQBuilder.ts
/*
 * This file is part of tswow (https://github.com/tswow)
 *
 * Copyright (C) 2020 tswow <https://github.com/tswow/>
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
import { ipaths } from '../util/Paths';
import { isWindows } from '../util/Platform';
import { wsys } from '../util/System';
import { bpaths, spaths } from './CompilePaths';

export namespace MPQBuilder {
    export async function create(cmake: string) {
        if (isWindows()) {
            wsys.exec(`${cmake} `
                + ` -S "${spaths.tools.mpqbuilder.get()}" `
                + ` -B "${bpaths.mpqbuilder.get()}"`
                + ` -DBOOST_ROOT="${bpaths.boost.boost_1_74_0.abs().get()}"`
                , 'inherit');
            wsys.exec(`${cmake}`
                + ` --build "${bpaths.mpqbuilder.get()}"`
                + ` --config Release`, 'inherit');
        } else {
            bpaths.mpqbuilder.mkdir()
            const relativeMpqSource = bpaths.mpqbuilder.relativeFrom(spaths.tools.mpqbuilder.get());
            await wsys.inDirectory(bpaths.mpqbuilder.get()
                , () => {
                    wsys.exec(
                        `${cmake} "${relativeMpqSource}"`
                        ,  'inherit');
                    wsys.exec(`make`,'inherit');
                });
        }
        bpaths.mpqbuilder.mpqbuilder_exe.copy(ipaths.bin.mpqbuilder.mpqbuilder_exe)
        bpaths.mpqbuilder.luaxml_exe.copy(ipaths.bin.mpqbuilder.luaxml_exe)
    }
}

File: mpqbuilder/mpqbuilder.cpp
/*
 * This file is part of tswow (https://github.com/tswow)
 * 
 * Copyright (C) 2020 tswow <https://github.com/tswow/>
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
#include <StormLib.h>
#include <boost/filesystem.hpp>
#include <boost/algorithm/string/predicate.hpp>
#include <iostream>
#include <functional>

namespace fs = boost::filesystem;

HANDLE handle = NULL;

void writeDir(fs::path basePath, fs::path curDir)
{
	fs::directory_iterator end;
	for (fs::directory_iterator itr(curDir); itr != end; ++itr)
	{
		if (fs::is_directory(itr->path()))
		{
			writeDir(basePath, itr->path());
		}
		else
		{
			auto relative = itr->path().string().substr(basePath.string().length(), itr->path().string().length());
			if (relative.rfind("/", 0) == 0 || relative.rfind("\\", 0) == 0)
			{
				relative = relative.substr(1, relative.length());
			}
			if (boost::algorithm::ends_with(relative, ".blend"))
			{
				return;
			}
			if (boost::algorithm::ends_with(relative, ".dbc"))
			{
				relative = "DBFilesclient/" + relative;
			}
			SFileAddFile(handle, itr->path().string().c_str(), relative.c_str(), 0);
		}
	}
}

int main(int argc, char **argv)
{
	if (argc < 3)
	{
		std::cout << "Usage: mpqbuilder outputFile inputDir1 (inputDir2...)";
		return -1;
	}

	auto outputFile = fs::path(argv[1]);
	if (!fs::exists((outputFile)))
	{
		if (!fs::remove((outputFile)))
		{
			std::cout << "Failed to remove old mpq file " << outputFile << "\n";
			return -1;
		}
	}

	if (fs::exists(outputFile))
	{
		if (!fs::remove((outputFile)))
		{
			std::cout << "Failed to remove old mpq file " << outputFile << "\n";
			return -1;
		}
	}

	if (!SFileCreateArchive(outputFile.string().c_str(), 0, 1024, &handle))
	{
		std::cout << "Failed to create output mpq file " << outputFile << "\n";
		return -1;
	}

	for (int i = 2; i < argc; ++i)
	{
		fs::path inputDir = fs::path(argv[i]);
		if (!fs::exists(fs::path(inputDir)))
		{
			std::cout << "No such directory " << inputDir << std::endl;
			return -1;
		}

		if (!fs::is_directory(fs::path(inputDir)))
		{
			std::cout << "Not a directory: " << inputDir << std::endl;
			return -1;
		}

		writeDir(inputDir, inputDir);
	}

	SFileFlushArchive(handle);
	return 0;
}